src/main/java/edu/harvard/iq/dataverse/validation/JSONDataValidation.java

```diff
@@ -0,0 +1,190 @@
+
+        if ("controlledVocabulary".equals(typeClass)) {
+            DatasetFieldType dsft = datasetFieldService.findByName(typeName);
+            if (value instanceof List<?>) {
+                ((List<?>) value).listIterator().forEachRemaining(cvv -> {
+                    if (datasetFieldService.findControlledVocabularyValueByDatasetFieldTypeAndStrValue(dsft, (String) cvv, true) == null) {
+                        throwValidationException("dataset", "cvv.missing", List.of(key, typeName, (String) cvv));
+                    }
+                });
+            } else {
+                if (datasetFieldService.findControlledVocabularyValueByDatasetFieldTypeAndStrValue(dsft, (String) value, true) == null) {
+                    throwValidationException("dataset", "cvv.missing", List.of(key, typeName, (String) value));
+                }
+            }
+        }
+    }
+
+    // If value is another object or list of objects that need to be further validated then childType refers to the parent
+    // Example: If this is a dsDescriptionValue from a dataset the messageType would be dataset.dsDescriptionValue
+    // This needs to match the Bundle.properties for mapping the error messages when an exception occurs
+    private static void validateChildren(Schema schema, Map<String, Map<String,List<String>>> schemaChildMap, String key, Collection<Object> children, String typeName, String messageType) {
+        if (children == null || children.isEmpty()) {
+            return;
+        }
+        List<String> requiredFields = new ArrayList<>();
+        requiredFields.addAll((List)schemaChildMap.getOrDefault(typeName, Collections.EMPTY_MAP).getOrDefault("required", Collections.EMPTY_LIST));
+        List<String> allowedFields = (List)schemaChildMap.getOrDefault(typeName, Collections.EMPTY_MAP).getOrDefault("allowed", Collections.EMPTY_LIST);
+        children.forEach(child -> {
+            if (child instanceof Map<?, ?>) {
+                String childTypeName = ((Map<?, ?>) child).containsKey("typeName") ? (String)((Map<?, ?>) child).get("typeName") : "";
+                if (!allowedFields.isEmpty() && !allowedFields.contains(childTypeName)) {
+                    throwValidationException(messageType, "invalidType", List.of(typeName, childTypeName, allowedFields.stream().collect(Collectors.joining(", "))));
+                }
+                if (!requiredFields.isEmpty() && requiredFields.contains(childTypeName)) {
+                    requiredFields.remove(childTypeName);
+                }
+            }
+        });
+        if (!requiredFields.isEmpty()) {
+            throwValidationException(messageType, "required.missing", List.of(typeName, requiredFields.stream().collect(Collectors.joining(", ")), typeName));
+        }
+    }
+    private static void throwValidationException(String key, List<String> argList) {
+        throw new ValidationException(BundleUtil.getStringFromBundle("schema.validation.exception." + key, argList));
```

Comment:
When I am trying to figure out why an error message is appearing I often search the code with the full bundle key implied by the error message. This kind of split, hardcoding only part of the key instead of the whole key makes that difficult.

Reply:
The key is dynamic. It could be a generic error message regardless of the specific field in question or it could be specific to a field.
ie. schema.validation.exception.value.missing is specific to the 'value' field where schema.validation.exception.dataset.required.missing is for any required field.

Reply:
No, I get why you are doing it this way - it's just that sometimes when I find myself debugging a production issue I will start with an error message or UI prompt and then find that verbiage in the bundle. Then I will search the code with the entire bundle key to see where in the code the message is coming from. When you split it like that the search will come up empty and I'll have to guess how much of the key to search on. Also some keys might end the same way. We can leave it like this, I'll just have to remind myself that there are times when I'll need to search with a partial bundle key.

Reply:
Yeah, I agree with this. I'm much more likely to search for the full `schema.validation.exception.value.missing` than the partial `value.missing` when hunting down a bug.

Raman used to do the same thing, used to not use the full key.

I can live with it for this PR but maybe we should write it up in our coding style page.